ci/infra/testrunner/tests/test_node_reboot.py

```diff
@@ -4,58 +4,44 @@
 
 import pytest
 
-from tests.utils import wait
+from tests.utils import (check_pods_ready, node_is_ready, wait)
 
 from timeout_decorator import timeout
 
 logger = logging.getLogger("testrunner")
 
 
-def check_pods_ready(kubectl):
-    pods = json.loads(kubectl.run_kubectl('get pods --all-namespaces -o json'))['items']
-    for pod in pods:
-        pod_status = pod['status']['phase']
-        pod_name   = pod['metadata']['name']
-        assert pod_status in ['Running', 'Completed'], f'Pod {pod_name} status {pod_status} != Running or Completed'
+def check_node_is_ready(platform, kubectl, role, nr):
+    assert node_is_ready(platform, kubectl, role, nr)
 
 
 @pytest.mark.pr
-@pytest.mark.parametrize('node_type,node_number', [('master', 1), ('worker', 0)])
-def test_hard_reboot(deployment, platform, skuba, kubectl, node_type, node_number):
-    assert skuba.num_of_nodes(node_type) > node_number
+@pytest.mark.parametrize('role,node', [('master', 1), ('worker', 0)])
+def test_hard_reboot(deployment, platform, skuba, kubectl, role, node):
+    """ Reboots master and worker nodes and checks the are back ready.
+    For masters, don't master 1, as master 0 is expected to be the
+    cluster leader and rebooting it may introduce transient etcd errors.
+    """
+    assert skuba.num_of_nodes(role) > node
 
-    logger.info('Wait for all the nodes to be ready')
-    wait(kubectl.run_kubectl, 
-         'wait --for=condition=Ready node --all --timeout=0',
-         wait_timeout=10,
-         wait_backoff=30,
-         wait_elapsed=300,
-         wait_allow=(RuntimeError))
+    platform.ssh_run(role, node, 'sudo reboot &')
 
-    logger.info(f'Rebooting {node_type} {node_number}')
-
-    platform.ssh_run(node_type, node_number, 'sudo reboot &')
-
-    # wait the node to reboot
-    wait(platform.ssh_run,
-         node_type,
-         node_number,
-         'echo hello',
-         wait_backoff=30,
-         wait_timeout=30,
-         wait_elapsed=300,
-         wait_allow=(RuntimeError))
+    # Allow time for the node to reboot and kubernetes to check readiness
+    time.sleep(60)
```

Comment:
It is not clear to me where do we wait for the node to boot. This sleep should be counted after the k8s daemon is running again. IMHO I'd wait for the node to be reachable again and then start counting the grace period.

Having an actively pulling call to check if the node is reachable can prevent failures if for some reason the node takes longer. We could have a longer timeout just for safety and it would not lead to longer run in most cases.

Reply:
We are waiting here for k8s to notice the node has gone not ready before we check it is ready. If we just check after rebooting the node, k8s may not have noticed it went not ready and the test will pass. However, the node can later go not ready and never come back ready.

Reply:
Correct, forget my previous comment. I did a wrong assumption that the verification call for readiness was done against the specific node, however the call is against the cluster.